teuthology/provision/maas.py

```diff
@@ -0,0 +1,430 @@
+                f"More than one machine found for hostname '{self.shortname}': {hostnames}"
+            )
+        return resp[0]
+
+    def get_image_data(self) -> Dict[str, Any]:
+        """Locate the image we want to use
+
+        :returns: The image data as a dictionary
+        """
+        resp = self.do_request("/boot-resources/").json()
+        if len(resp) == 0:
+            raise RuntimeError("MaaS has no images available")
+
+        os_version = OS._version_to_codename(self.os_type, self.os_version)
+        name = f"{self.os_type}/{os_version}"
+        for image in resp:
+            if image["name"] == name:
+                return image
+        raise RuntimeError(f"MaaS has no {name} image available")
+
+    def lock_machine(self) -> None:
+        """Lock the machine"""
+        resp = self.do_request(
+            f"/machines/{self.system_id}/op-lock", method="POST"
+        )
+
+        if resp.text == "Machine is locked":
+            self.log.info(f"Machine '{self.shortname}' is locked")
+        elif data := resp.json():
+            if not data.get("locked"):
+                raise RuntimeError(
+                    f"Machine '{self.shortname}' locking failed, "
+                    f"Current status: {data.get('locked')}"
+                )
+
+    def unlock_machine(self) -> None:
+        """Unlock the machine"""
+        resp = self.do_request(
+            f"/machines/{self.system_id}/op-unlock", method="POST"
+        )
+
+        if resp.text == "Cannot unlock an already-unlocked node":
+            self.log.info(f"Machine '{self.shortname}' is not locked")
+        elif data := resp.json():
+            if data.get("locked"):
+                raise RuntimeError(
+                    f"Machine '{self.shortname}' locking failed, "
+                    f"Current status: {data.get('locked')}"
+                )
+
+    def deploy_machine(self) -> None:
+        """Deploy the machine"""
+        image_data: Dict[str, Any] = self.get_image_data()
+        if image_data.get("type", "").lower() not in ["synced", "uploaded"]:
+            raise RuntimeError(
+                f"MaaS image {image_data.get('name')} is not synced, "
+                f"current status: {image_data.get('type')}"
+            )
+
+        self.log.info(
+            f"Deploying machine '{self.shortname}' with image {image_data.get('name')}"
+        )
+        files = {
+            "distro_series": (None, image_data.get("name")),
+            "user_data": (None, self._get_user_data()),
+        }
+        data: Dict[str, Any] = self.do_request(
+            f"/machines/{self.system_id}/op-deploy", method="POST", files=files
+        ).json()
+        if data.get("status_name", "").lower() != "deploying":
+            raise RuntimeError(
+                f"Machine '{self.shortname}' deployment failed, "
+                f"Current status: {data.get('status_name')}",
+            )
+
+    def release_machine(self, erase: bool = True) -> None:
```

Comment:
is erase = True really the right default?  It'll take more time, and given that we're redeploying the OS, does it accomplish anything?